helix-view/src/document.rs

```diff
@@ -163,26 +174,128 @@ impl Document {
         }
     }
 
+    #[inline(always)]
+    fn calculate_decode_capacity(decoder: &mut encoding_rs::Decoder, buf: &[u8]) -> usize {
+        // Will never panic because the buffer size will never be close to usize::MAX in the
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = decoder
+            .max_utf8_buffer_length(buf.len() - valid_up_to)
+            .unwrap();
+        valid_up_to + max_buffer_len
+    }
+
+    #[inline(always)]
+    fn calculate_encode_capacity(encoder: &mut encoding_rs::Encoder, buf: &[u8]) -> usize {
+        // Will never panic because the buffer size will never be close to usize::MAX in the
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = encoder
+            .max_buffer_length_from_utf8_if_no_unmappables(buf.len() - valid_up_to)
+            .unwrap();
+        (valid_up_to + max_buffer_len).next_power_of_two()
+    }
+
     // TODO: async fn?
-    pub fn load(path: PathBuf) -> Result<Self, Error> {
-        use std::{fs::File, io::BufReader};
+    /// Open a file from a path, or create a new file if it doesn't exist.
+    fn _open(
+        path: &PathBuf,
+        encoding: Option<&'static encoding_rs::Encoding>,
+    ) -> Result<(Rope, &'static encoding_rs::Encoding), Error> {
+        use std::{
+            fs::File,
+            io::{BufReader, Read},
+        };
 
-        let doc = if !path.exists() {
-            Rope::from("\n")
-        } else {
-            let file = File::open(&path).context(format!("unable to open {:?}", path))?;
-            let mut doc = Rope::from_reader(BufReader::new(file))?;
-            // add missing newline at the end of file
-            if doc.len_bytes() == 0 || doc.byte(doc.len_bytes() - 1) != b'\n' {
-                doc.insert_char(doc.len_chars(), '\n');
-            }
-            doc
+        if !path.exists() {
+            return Err(anyhow::anyhow!("path doesn't exist"));
+        }
+
+        // Create a 64kb buffer for reading the file into.
+        let mut buf = [0u8; 1 << 16];
+        let mut builder = RopeBuilder::new();
+        let mut file = BufReader::new(File::open(&path)?);
```

Comment:
Will not using `BufReader` here makes it faster since we have some sort of buffer already.

Reply:
Ah, that's a good point.  Yeah, since we're doing our own buffering, `BufReader` is redundant and probably costing a little bit of performance.